realm/realm-library/src/objectServer/java/io/realm/PermissionManager.java

```diff
@@ -1090,6 +1094,69 @@ private void notifyCallbackWithSuccess(RealmResults<PermissionOffer> permissions
         }
     }
 
+    private class RevokeOfferAsyncTask extends PermissionManagerTask<Permission> {
+
+        private final String offerToken;
+        private final RevokeOfferCallback callback;
+        private RealmResults<PermissionOffer> matchingOffers;
+
+        public RevokeOfferAsyncTask(PermissionManager permissionManager, String offerToken, RevokeOfferCallback callback) {
+            super(permissionManager, callback);
+            this.offerToken = offerToken;
+            this.callback = callback;
+        }
+
+        @Override
+        public void run() {
+            if (checkAndReportInvalidState()) {
+                return;
+            }
+            matchingOffers  = managementRealm.where(PermissionOffer.class)
+                    .equalTo("token", offerToken)
+                    .findAllAsync();
+            matchingOffers.addChangeListener(new RealmChangeListener<RealmResults<PermissionOffer>>() {
+                @Override
+                public void onChange(final RealmResults<PermissionOffer> offers) {
+                    if (offers.size() == 1) {
+                        managementRealm.executeTransactionAsync(new Realm.Transaction() {
+                            @Override
+                            public void execute(Realm realm) {
+                                // Make 100% sure the offer is still in the DB.
+                                // It could have been deleted between querying for it and the
+                                // transaction running. We will still call OnSuccess if the
+                                // offer was removed by someone else.
+                                PermissionOffer offer = realm.where(PermissionOffer.class)
+                                        .equalTo("token", offerToken)
+                                        .findFirst();
+                                if (offer != null) {
+                                    RealmObject.deleteFromRealm(offer);
+                                }
+                            }
+                        }, new Realm.Transaction.OnSuccess() {
+                            @Override
+                            public void onSuccess() {
+                                matchingOffers.removeAllChangeListeners();
+                                notifyCallbackWithSuccess();
+                            }
+                        }, new Realm.Transaction.OnError() {
+                            @Override
+                            public void onError(Throwable error) {
+                                matchingOffers.removeAllChangeListeners();
+                                notifyCallbackError(new ObjectServerError(ErrorCode.UNKNOWN, error));
+
+                            }
+                        });
+                    }
+                }
+            });
+        }
+
+        private void notifyCallbackWithSuccess() {
```

Comment:
I actually just saw something very similar in `SQLBrite` today. see [here](https://github.com/square/sqlbrite/blob/4d4969e5840d330c413034fbf83b5e7e9e86743f/sqlbrite/src/main/java/com/squareup/sqlbrite2/BriteDatabase.java#L69)